core/src/main/scala/chisel3/Layer.scala

```diff
@@ -40,13 +26,7 @@ object layer {
     * @param outputDirBehavior an optional user-provided output directory for this layer
     * @param _parent the parent layer, if any
     */
-  abstract class Layer(
-    val convention:        Convention.Type,
-    val outputDirBehavior: OutputDirBehavior = DefaultOutputDir
-  )(
-    implicit _parent: Layer,
-    _sourceInfo:      SourceInfo) {
-    self: Singleton =>
+  abstract sealed class Layer(implicit _parent: Layer, _sourceInfo: SourceInfo) {
```

Comment:
Scala nit: use of `abstract` on the three classes in this diff isn't accurate as there are no abstract members. You _can_ do this in Scala if you want to prevent a user from doing `new Layer` and force them to do `new Layer {}`. However, I don't think there is much benefit in that---particularly for the `BindLayer` and `InlineLayer` which are self-typed with `Singleton` which I think mandates that you can't construct them via `new`.

Reply:
This change is a hard API breakage. We can do this. It will just break code and should be communicated as such.